Ensure that widgets can be garbage collected on each platform

samschott opened the pull request:
This WIP PR starts by adding tests for the garbage collection of widgets. Known issues causing test failures are:

* Circular references between winforms and Python. Will be fixed by #2066.

Other test failures will need to be investigated.

## PR Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

freakboy3742 commented on testbed/tests/widgets/test_box.py:
```diff
@@ -15,3 +18,13 @@
     return toga.Box(style=Pack(width=100, height=200))
+
+
+async def test_cleanup():
```
Given this test is identical on every widget, it should be possible to pull this test definition into a generic test in `properties.py`, import it into the module, and use the `widget` fixture to provide the actual widget instance. This is how tests like `test_focus_noop` work; AFAICT the same pattern should work here.

samschott replied:
Potentially. Getting `widget` from a pytest fixture means that it will be cached until cleanup of the fixture's scope, which is always after a test case.

I haven't really found a good way to instantiate the widget in the test case itself but still share the test logic. Rewriting the fixture to return a "widget factory" might be an option, but will harm readability of all test cases. See https://docs.pytest.org/en/stable/how-to/fixtures.html#factories-as-fixtures.

samschott replied:
I've at least moved most of the shared code to `assertions.py`.

However, there seems to be no easy way to work around fixture magic to ensure that there are other references to the widget. Even if there was, I'd be afraid if it being brittle and breaking with pytest changes, so I'd prefer to explicitly initialize the widget as part of the test cases.

samschott replied:
Hrmp, even that does not seem to work and still leaves references in Python 3.10 (though it passes on Python 3.12). Reverted for now.

freakboy3742 replied:
We might not be able to use fixtures, but we could still use a factory function so that the logic of the test isn't replicated:

```
def _cleanup_test(Widget, args=None, kwargs=None):
    async def test_cleanup():
        widget = Widget(
            *(args if args is not None else ()),
            **(kwargs if kwargs is not None else {}),
        )
        ref = weakref.ref(widget)

        del widget
        gc.collect()

        assert ref() is None

    return test_cleanup
```
adding in the extra logic for a generic skip and xfail as well; usage is then:
```
test_cleanup = _cleanup_test(Box)
```

samschott replied:
Fair point, something like this should work. I've run into subtle issues with this for "container" widgets where content widgets are passed as arguments.

Generally, those will hold a backref to the container and therefore need to be cleanup of as well before the assertion, instead of after the `args` and `kwargs` go out of scope.

On Windows, there seems to be an additional issue with those argument widgets being created on a different thread than where the widget under test is created. I'll see if this can be solved somehow without passing a full `create_widget_under_test` function to the test case.

samschott commented:
I had completely forgotten about this PR, thanks for the reminder! Cleaned it up a bit:

1. Added missing tests to some new widgets.
2. Added xfails to widgets / platforms with memory leaks. I chosen xfail over skip for now, to ensure that I don't accidentally add superfluous skips, and that they get cleaned up whenever leaks are fixed "accidentally".

freakboy3742 reviewed (approved):
Looks good. There's obviously a follow up task to actually address the memory leaks; I'll open a separate ticket to track this.
